text/3185-static-async-fn-in-trait.md

```diff
@@ -0,0 +1,240 @@
+# Draft RFC: Static async fn in traits
+
+- Feature Name: `async_fn_in_traits`
+- Start Date: 2021-10-13
+- RFC PR: [rust-lang/rfcs#3185](https://github.com/rust-lang/rfcs/pull/3185)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Support `async fn` in traits that can be called via static dispatch. These will desugar to an anonymous associated type.
+
+# Motivation
+[motivation]: #motivation
+
+Async/await allows users to write asynchronous code much easier than they could before. However, it doesn't play nice with other core language features that make Rust the great language it is, like traits.
+
+In this RFC we will begin the process of integrating these two features and smoothing over a wrinkle that async Rust users have been working around since async/await stabilized nearly 3 years ago.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+You can write `async fn` in traits and trait impls. For example:
+
+```rust
+trait Service {
+    async fn request(&self, key: i32) -> Response;
+}
+
+struct MyService {
+    db: Database
+}
+
+impl Service for MyService {
+    async fn request(&self, key: i32) -> Response {
+        Response {
+            contents: self.db.query(key).await.to_string()
+        }
+    }
+}
+```
+
+This is useful for writing generic async code.
+
+Currently, if you use an `async fn` in a trait, that trait is not `dyn` safe. If you need to use dynamic dispatch combined with async functions, you can use the [`async-trait`] crate. We expect to extend the language to support this use case in the future.
+
+Note that if a function in a trait is written as an `async fn`, it must also be written as an `async fn` in your implementation of that trait. With the above trait, you could not write this:
```

Comment:
Is there a possibility that this restriction could be later lifted in a backwards-compatible way?

This is important, because with the explicit `impl Future` you're able to run some preliminary logic before constructing the final future, which makes it possible for the future _not_ to capture some of the input arguments, potentially making the future `Send` where it wouldn't be otherwise.

In regular async functions, it's sort of an implementation detail of the function whether it's implemented as an actual `async fn` or as an `fn -> impl Future` — while this fact "leaks" to rustdoc and elsewhere, it can be more or less flipped back and forth without breaking the public interface of the function. Whereas this design would "bake in", at the trait level, the fact that implementation functions _must_ use the `async fn` syntax, even though the users of the trait don't actually care since they invoke a method and get a future from it either way.

So my concern is, _would_ anyone actually use the `async fn` syntax when defining a trait, knowing that this forces an unnecessary requirement on all implementations?

Reply:
@bugaevc IMHO, as you have already alluded to, these function signatures are not interchangeable and I don't think they should be treated as such. At a glance `async fn` tells me that it is fully async and not `await`ing has no side effects at all. `fn -> impl Future` on the other had may or may not have side effects. Not `await`ing it may also block the thread or not. To cut the long story short, I think these function signatures provide different guarantees. A fact we might want to make explicit.

Reply:
One way that this could happen would be for `impl`s to infer the associated type's values.

That would be nice for things other than `async fn` too -- one could imagine
```rust
impl Iterator for Foo {
    fn next(&mut self) -> Option<i32> { ... }
}
```
just working without needing to say `type Output = i32;`, treating the associated type kinda like it was a TAIT.

(With TAIT it could be written something like 
```rust
type BlahBlah = impl Any;
impl Iterator for Foo {
    type Output = BlahBlah;
    fn next(&mut self) -> Option<i32> { ... }
}
```
which is noisier without really adding anything particularly useful to the reader.)

Reply:
Yes, I had the same idea. Inferring associated types would be very nice outside of async traits as well. The downside is that when reading a trait impl, you might have to do a bit more searching to identify a given asociated type.

Reply:
We can allow this later by adding return position `impl Trait` in traits and making this feature interoperable with it (i.e., redefine the `async fn` in traits desugaring in terms of return position `impl Trait`). I think we should do so, because as you and @cramertj point out, library authors will want to know they aren't backing themselves into a corner by using this feature.